Training/Program.cs

```diff
@@ -1,7 +1,102 @@
    internal class Program {
-      static void Main (string[] args) {
-         Console.WriteLine ("Hello, World!");
+      static void Main () { }
+   }
+   #region Queue ------------------------------------------------------------------------------
+
+   public class Deque<T> {
+      #region Properties --------------------------------------------
+      /// <summary>Capacity of the TDeque</summary>
+      public int Capacity => mElements.Length;
+
+      /// <summary>Count of the elements in TDeque</summary>
+      public int Count => mCount;
+
+      /// <summary>Tells if the queue is empty or not</summary>
+      public bool IsEmpty => mCount == 0;
+      #endregion
+
+      #region Method ------------------------------------------------
+      // Throws exception when queue is empty
+      void CheckEmpty () {
+         if (IsEmpty) throw new InvalidOperationException ("Empty Queue");
+      }
+
+      /// <summary>Adds an element at front of the queue</summary>
+      /// <param name="a">Elements of queue</param>
+      public void EnqueueFront (T a) {
+         if (mCount == mElements.Length) Resize ();
+         // Move the pointer to front position
+         mFront = (mFront - 1 + mElements.Length) % mElements.Length;
+         mElements[mFront] = a;
+         mCount++;
+      }
+
+      /// <summary>Adds an element at end of the queue</summary>
+      /// <param name="a">Elements of queue</param>
+      public void EnqueueRear (T a) {
+         if (mCount == mElements.Length) Resize ();
+         mElements[mRear] = a;
+         mRear = (mRear + 1) % mElements.Length;
+         mCount++;
+      }
+
+      /// <summary>Removes an element at end of the queue</summary>
+      /// <returns>Returns removed element</returns>
+      public T DequeueRear () {
+         CheckEmpty ();
+         // Move the pointer to rear position
+         mRear = (mRear - 1 + mElements.Length) % mElements.Length;
+         mCount--;
+         return mElements[mRear];
+      }
+
+      /// <summary>Removes an element at front of the queue</summary>
+      /// <returns>Returns removed element</returns>
+      public T DequeueFront () {
+         CheckEmpty ();
+         (T a, mFront) = (mElements[mFront], (mFront + 1) % mElements.Length);
+         mCount--;
+         return a;
+      }
+
+      /// <summary>Returns the first element in queue without removing it</summary> 
+      public T PeekFront () {
+         CheckEmpty ();
+         return mElements[mFront];
+      }
+
+      /// <summary>Returns the last element in queue without removing it</summary> 
+      public T PeekRear () {
+         CheckEmpty ();
+         return mElements[(mRear - 1 + mElements.Length) % mElements.Length];
+      }
+
+      // Double the capacity of queue when needed
+      void Resize () {
+         var temp = new T[Capacity * 2];
+         for (int i = 0; i < Capacity; i++) {
+            temp[i] = mElements[mRear];
+            mRear = (mRear + 1) % Capacity;

```

Comment:
You could use i and mRear, instead of incrementing mRear vaariable. Don't change mRear variable.